### DISCUSSION TITLE
Data and error from the same object

### DISCUSSION THREAD

**Discussion by arnabkd** [graphql-java/graphql-java#2642] [Q&A] [answered]
Say I have a class `Foo` that looks like this (Kotlin code):

```
class Foo {
  fun ok(): Boolean = true
  fun notOk(): Boolean = throw IllegalArgumentException("test")
}
```

Now say I have a schema where I define the following:
```
type Query {
  getFoo: Foo!
}

type Foo {
  ok: Boolean!
  notOk: Boolean!
}
```

Assume that `getFoo` is in a resolver somewhere. Now if I do the following query:
```
{
  getFoo {
    ok
    notOk
  }
}
```

The whole thing just fails (as expected), but what I want to achieve is that *only* the `notOk` field should turn up in the `error` part of the response and the `ok` field should turn up as part of the `data` part of the response.

For reference, I am using: https://github.com/excitement-engineer/ktor-graphql

As far as I can understand, I need to make a `DataFetcherExceptionHandler` object, but I don't understand how I can both fulfill the non-erroring fields and the erroring field. Any ideas? This is what I have so far:
```
val exceptionHandler = DataFetcherExceptionHandler { handlerParameters ->  
  val exception = handlerParameters.exception
  val sourceLocation = handlerParameters.sourceLocation
  val path = handlerParameters.path

  val error = object : GraphQLError {
    override fun getMessage() = exception.message
    override fun getLocations() = mutableListOf(sourceLocation)
    override fun getErrorType() = ErrorType.DataFetchingException
  }

  DataFetcherExceptionHandlerResult.newResult().error(error).build()
}
```

**Comment by learnerEvermore** [comment, marked as answer]
Have a look at the `DataFetcher`:

``` java
/**
 * A data fetcher is responsible for returning a data value back for a given graphql field.  The graphql engine
 * uses data fetchers to resolve / fetch a logical field into a runtime object that will be sent back as part
 * of the overall graphql {@link graphql.ExecutionResult}
 *
 * In other implementations, these are sometimes called "Resolvers" or "Field Resolvers", because that is there function,
 * they resolve a logical graphql field into an actual data value.
 *
 * @param <T> the type of object returned. May also be wrapped in a {@link graphql.execution.DataFetcherResult}
 */
@PublicSpi
public interface DataFetcher<T> {

    /**
     * This is called by the graphql engine to fetch the value.  The {@link graphql.schema.DataFetchingEnvironment} is a composite
     * context object that tells you all you need to know about how to fetch a data value in graphql type terms.
     *
     * @param environment this is the data fetching environment which contains all the context you need to fetch a value
     *
     * @return a value of type T. May be wrapped in a {@link graphql.execution.DataFetcherResult}
     *
     * @throws Exception to relieve the implementations from having to wrap checked exceptions. Any exception thrown
     *                   from a {@code DataFetcher} will eventually be handled by the registered {@link graphql.execution.DataFetcherExceptionHandler}
     *                   and the related field will have a value of {@code null} in the result.
     */
    T get(DataFetchingEnvironment environment) throws Exception;
}
```

Notice that it can return anything really and, regardless of `<T>` it can also return a `DataFetcherResult` instead, wrapping the value.  Have a look at it. It can contain both `data` and `errors` - i.e. you can call both `data(...)` and `error(...)` on it to set both.

**Reply by arnabkd** [reply]
But does this support having a class `Foo` with one erroring field and one successfully resolving field? It seems to me that this would be needed to be implemented in a special way per field that can throw an error, no?

So in my example it would have to be a class `Foo` that again implements a data fetcher for the field `baz` (of type `Boolean`), and that would actually be of type `Boolean?`. 

Would that mean that I can resolve both `baz` and `bar` fields inside of the same instance of `Foo` (one errors and the other does not)?

**Comment by learnerEvermore** [comment]
You can have as many errors (0-n) and optionally data (0-1) for any field, at any nesting level. Implementor is responsible for providing them and, in my experience, `graphql-java` does to the job of including them in the response. You can have literally:

``` graphql
{
    weird {
        completelyOK
        notAtAllOK
        partiallyOK
    }
}
```
... and have errors returned for `notAtAllOK` and `partiallyOK` and (some) data returned for `completelyOK` and `partiallyOK`.

**Reply by arnabkd** [reply]
It seems that the only thing I actually needed to do was mark the field as `Boolean?` in Kotlin and `Boolean` in the schema. Apparently `ktor-graphql` does this for me when I have a nullable field that might result in an error/exception being thrown. I still have the exception handler to avoid leaking implementation details to the client, but that's a separate concern. Thanks for the reply!

**Comment by sshevlyagin** [comment]
Resurrecting an old thread because I'm having a similar conversation. While you definitely can have a field that returns both data and has an error, that goes against the [spec](https://spec.graphql.org/October2021/#sel-FANRNHCAACEWwsN) with some good discussion in https://productionreadygraphql.com/2020-08-01-guide-to-graphql-errors in the ❗️Con: Exceptional Only section.

@arnabkd the error should be modeled in the schema if it's an error state that's part of the API contract, for example user name already taken when someone is trying to create a new account. If it's an exceptional developer error (the load balancer is down), then the field should just be null and there should be an error in the array.

**Reply by bbakerman** [reply]
Now that is interesting... and new.  This was not specified in previous specs

https://spec.graphql.org/October2021/#sec-Handling-Field-Errors

vs say

https://spec.graphql.org/June2018/#sec-Errors-and-Non-Nullability

**Reply by sshevlyagin** [reply]
What do you mean @bbakerman, the two sections are basically the same. 

**[2021](https://spec.graphql.org/October2021/#sec-Handling-Field-Errors)**

> If a field error is raised while resolving a field, it is handled as though the field returned null, and the error must be added to the "errors" list in the response.
> If the result of resolving a field is null (either because the function to resolve the field returned null or because a field error was raised), and that field is of a Non-Null type, then a field error is raised. The error must be added to the "errors" list in the response.

**[2018](https://spec.graphql.org/June2018/#sec-Errors-and-Non-Nullability)**
> If an error is thrown while resolving a field, it should be treated as though the field returned null, and an error must be added to the "errors" list in the response.
> If the result of resolving a field is null (either because the function to resolve the field returned null or because an error occurred), and that field is of a Non-Null type, then a field error is thrown. The error must be added to the "errors" list in the response.
